dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -3579,21 +3608,28 @@ RtpsUdpDataLink::send_heartbeats(const DCPS::MonotonicTimePoint& /*now*/)
     }
   }
 
-  if (writers_.empty() && interesting_readers_.empty()) {
-    heartbeat_.disable_and_wait();
+  if (expected_acks_) {
+    // One more readers did not ack since the last heartbeat.
+    heartbeat_period_ *= cfg.heartbeat_backoff_factor_;
+  } else if (last_ack_ != MonotonicTimePoint::zero_value && last_heartbeat_ != MonotonicTimePoint::zero_value) {
+    heartbeat_period_ = (last_ack_ - last_heartbeat_) * cfg.heartbeat_safety_factor_;
   }
+  heartbeat_period_ = std::min(heartbeat_period_, cfg.heartbeat_period_maximum_);
+  heartbeat_period_ = std::max(heartbeat_period_, cfg.heartbeat_period_minimum_);
+  expected_acks_ = 0;
 
   using namespace OpenDDS::RTPS;
 
   MetaSubmessageVec meta_submessages;
 
+  bool keep_going = false;
```

Comment:
Should `keep_going` impact the `for` loop two lines down?

Reply:
I've pushed code that initializes `keep_going`.  Does that answer your question?